Updates the Blaze press widget flow to handle a cleanup step

sbarbosa opened the pull request:
I am working on integrating Sentry.io into the Promote post widget. Because the widget is an external standalone app that runs in the same browser window, I must be careful about conflicting with the existing Calypso Sentry.io configuration.

After discussing the best approach with the Calypso core team, we decided to share the same Sentry instance as Calypso. We have to consider the difference in traffic that the widget gets, in comparison with Calypso, so we will be applying some changes to the calypso-sentry package which will allow us to enable Sentry for a wider audience when they open the DSP widget.

We will need to revert that operation after the users finish promoting a post, so we will be adding a cleanup step to the Promote widget that we will run each time the modal is closed. 

## Proposed Changes

* Adds the cleanup logic to the Promote post widget
* Modify the calypso-sentry package to allow us to re-initialize it with a different sampleRate
* Modify the show/close widget entry points to initialize and close Sentry so the widget can report all errors.

In another parallel PR, the widget will deploy a new version that will append a new eventPreprocessor to Sentry, which will tag errors originating from the widget with a custom tag and release. This will allow us to create personalized alerts for the Promote post widget.

## Reviewing Notes

You can ignore the changes of the files:
```
config/wpcalypso.json
client/my-sites/promote-post-i2/main.tsx
```

I Changed them to ease the testing process, but I will revert them before merging the PR.


## Testing Instructions
_edit: we now keep Sentry enabled after closing the widget, so the last few steps don't apply._

**Scenario 1: User without Sentry enabled, opens the widget**

* Open the webview and navigate to the Advertising section (Tools->Advertising)
* Open your browser console
* Append `?flags=catch-js-errors` to the URL and refresh the page. Verify that the message `Initializing error reporting...` is not reported in the console. (if you see it, refresh the page again)
* Click on the Promote button on any of your posts
* Verify that in the console, the reporting initialization text is shown. (We enable Sentry for all of the users that opens the widget.)
* Close the widget and then click on the Promote button at the top of the page, to generate an error
* In the Sentry administrative page for the calypso project, verify that the error we just generated was not reported.

**Note:** If the user doesn't have Sentry enabled, we leave it disabled when the user finishes working on the widget.

**Scenario 2: User with Sentry enabled, opens the widget**

* In the Advertising page, append `?flags=always-enable-sentry` to the URL to force Sentry initialization. When the page loads, verify that you see the message `Initializing error reporting...`
* Click on the Promote button for any of your posts
* Close the widget and then click on the Promote button at the top of the page, to generate an error
* In the Sentry administrative page for the calypso project, verify that the error we just generated was reported.

**Note:** If the user has Sentry enabled, we leave it enabled when the user finishes working on the widget.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

sbarbosa committed:
Updates the Blaze press widget flow to handle a cleanup step

sbarbosa committed:
Improves cleanup mechanism

sbarbosa committed:
Reverts previos unnecessary change

sbarbosa committed:
Merge branch 'trunk' into update/promote-post-widget-cleanup-step

sbarbosa committed:
Changes to share Sentry instance between the DSP widget and Calypso

sbarbosa committed:
Merge branch 'trunk' into update/promote-post-widget-cleanup-step

sbarbosa committed:
Added additional reporting to the widget init flow

noahtallen commented on config/wpcalypso.json:
```diff
@@ -10,7 +10,7 @@
 	"facebook_api_key": "249643311490",
 	"dsp_stripe_pub_key": "pk_live_51LYYzQF53KN4RFN0oHFYHpzmHGpuUAXZ6ygDAFxM4XkRArRsET0orrofmsynkd9DhFaOGEsDZ3RC4f8mgI0zhEyN000X8i0Mqx",
-	"dsp_widget_js_src": "https://dsp.wp.com/widget.js",
+	"dsp_widget_js_src": "https://widgets.wp.com/promote-v2/widget.js",
```
I'm not super familiar with this DSP stuff, but I did want to mention that we have other scripts developed in calypso `e.g. /apps/notifications` that are both deployed to widegts.wp.com and also imported normally in Calypso.

noahtallen commented on config/wpcalypso.json:
```diff
@@ -24,7 +24,7 @@
 		"calypsoify/plugins": true,
 		"cancellation-offers": true,
-		"catch-js-errors": false,
+		"catch-js-errors": true,
```
By the way, you can use `?flags=always-enable-sentry` in the URL to bypass the check which disables error reporting in development

sbarbosa replied:
Wow, I totally forgot that! You are right!

noahtallen commented on packages/calypso-sentry/src/index.ts:
```diff
@@ -118,36 +118,63 @@ function beforeBreadcrumb( breadcrumb: SentryApi.Breadcrumb ): SentryApi.Breadcr
+			( state.state === 'initial' && ! parameters ) ||
+			! [ 'initial', 'disabled' ].includes( state.state )
+		) {
+			return 'enabled'; // Previous state: enabled -> We're already running [loading, loaded, error]
```
This isn't completely accurate -- for example, sentry should never be enabled if `typeof document === 'undefined'`. That should probably be a different if check

noahtallen replied:
Sorry for contradicting myself. :) But now I understand it better, i think all these lines could be simplified to:

```ts
// No Sentry loading on the server.
// No double-loading. (Only continue if the state is initial or disabled to let the caller enable Sentry if it was previously disabled.)
if ( typeof document === 'undefined' || ! [ 'initial', 'disabled' ].includes( state.state ) ) {
	return;
}
```

noahtallen reviewed (changes requested):
Sorry for taking so long to review this!

Thinking about this problem again, I think we could vastly simplify the PR. Our problem currently is that Sentry is async loaded 10% of the time, meaning it won't be available for this widget. Essentially, when the widget loads, we need to make sure that Sentry is available.

There are basically two scenarios:
1. Sentry already loaded. In this case, the sample rate is already 100%, and the trunk code works fine.
2. Sentry is not loaded, in which case we just need to make sure it _does_ load when it's needed for this widget. (And then it's fine for Sentry to stay available after that, because we already loaded the script in the browser.)

So IMO we just need a function like this:

```js
function forceLoadSentry() {
  if ( state === 'disabled' ) {
    initSentry( { forceLoad: true } );
  }
}
``` 

Then we just make a few small tweaks to `initSentry`, with the goal of triggering the async load again. Just need to make sure we don't exit early again when forceLoad is true:
- If `forceLoad` is true, don't return early from the double-load check.
- If `forceLoad` is true, don't return early from the `shouldEnableSentry` check
- Persist the original options passed to `initSentry` so that they can be re-used if unavailable. Could just save `beforeSend` and `userId` to a global the first time `initSentry` is called.

Then when it's needed for DSP, just call `forceLoadSentry` -- that'd make sure that Sentry is available only if it was previously disabled. And you can call `forceLoadSentry` as many times as you want, and it will normally not try to load anything twice due to the disabled check.

I don't think closing sentry afterwards is necessary -- the only reason we conditionally disable it is to avoid loading more large scripts. Since `forceLoadSentry` will already load the script, we might as well continue using the instance.

I think with this approach, we could avoid almost all of the extra steps we have in the PR currently.

And it should still be possible to do the event processing. In theory, you _could_ pass your own params through `forceLoadSentry`, but I don't think it's necessary.

sbarbosa commented:
Hi @noahtallen! Thanks for the review.

The solution that you mentioned here is very similar to the solution I tried to implement. It is true that if we don't care about closing Sentry, then the code becomes even more straightforward. 

I'd like to tell you why I couldn't use some of what you proposed here.

Related to the forceLoadSentry function:
```
function forceLoadSentry() {
  if ( state === 'disabled' ) {
    initSentry( { forceLoad: true } );
  }
}
```
I first tried this approach, but the main problem here is that Calypso needs to have the beforeSend function when we initialize Sentry. Currently, it is initiated at [boot time](https://github.com/Automattic/wp-calypso/blob/0f54920194f6ea8d0ef469553af4c32a49040d1c/client/boot/common.js#L244-L262), and at that time, we have the redux store in our context. 

I analyzed a way to export and re-use that function, but because it uses the hole redux store, it won't be possible when the widget starts. And taking into account how redux works, I am not sure we can generate the same function using a connected component here. (I fear the state we used in the beforeSend function won't have the latest change we made to the store).

With that limitation, I took the approach of storing the initial parameters we send to Sentry in the library to be re-used in future initialization. 

Another point is that I wanted to leave the sample rate open and configurable. We want to log every user that opens the widget right now, but if we start getting more load there, we will reduce the amount. That is the reason why I added a sampleRate parameter.

---

Those were the main reasons why the final implementation ended up like that. Maybe you can find a better way to persist/recreate the beforeSend function so we can send it when we call the initSentry. 

The changes I applied to the library:
1- Persist in the state of the previous execution parameters. 
I did this to have the beforeSend function at hand. With this addition, we can call the initSentryFunction anywhere without changing how it was implemented (we have some unity).
2- Added a sampleRate parameter.
This was mostly thinking about the future. We will need to adapt the percentage when we start to see an increase in traffic.
3- Added a return parameter to the initSentry function with the previous state
This was needed to maintain the status of Sentry when we closed the widget. We can improve this and remove it if you think this won't increase too much the amount of reported errors

Let me know if you can think of better ways to handle point 1. And your opinions about points 2 and 3.

sbarbosa committed:
Merge branch 'trunk' into update/promote-post-widget-cleanup-step

sbarbosa committed:
Improves server check condition

noahtallen commented on packages/calypso-sentry/src/index.ts:
```diff
@@ -118,36 +118,66 @@ function beforeBreadcrumb( breadcrumb: SentryApi.Breadcrumb ): SentryApi.Breadcr
+		const params =
+			state.state === 'initial'
+				? ( parameters as SentryOptions ) // We know it's not undefined because of the check above
+				: { ...state.params, sampleRate: 0.1, ...parameters };
```
We don't need to store a default sampleRate in state.params since we set a default value. IMO, we can put the default `sampleRate` value into the `shouldEnableSentry` function so that it's closer to where it's used

noahtallen commented on packages/calypso-sentry/src/index.ts:
```diff
@@ -118,36 +118,66 @@ function beforeBreadcrumb( breadcrumb: SentryApi.Breadcrumb ): SentryApi.Breadcr
-			return;
+		// No initializing with empty params
+		if (
+			( state.state === 'initial' && ! parameters ) ||
```
Since every param is optional, I don't think this is necessary

noahtallen commented on packages/calypso-sentry/src/index.ts:
```diff
@@ -218,4 +248,16 @@ export async function initSentry( { beforeSend, userId }: SentryOptions ) {
 		clearQueues();
 	}
+
+	return 'disabled'; // Previous state: disabled -> We were not running [initial, disabled]
```
If we go with the approach I mentioned, we can remove all extra return statements.

noahtallen reviewed (commented):
> I did this to have the beforeSend function at hand. With this addition, we can call the initSentryFunction anywhere without changing how it was implemented

👍 This makes sense to me.

> Added a sampleRate parameter.

Hm, I see what you mean. I think this could get confusing if more people go with this approach. Ideally, we would just rely on the global sample rate or a custom sampler function to simplify things. For example, the custom sampler function could capture a lower % of exceptions from this part of the code for anyone that has Sentry loaded. 

But I'm ok keeping it for now.

> Added a return parameter to the initSentry function with the previous state

I think we should remove this. As I mentioned, I really don't think we need to cleanup/restore Sentry after using it for DSP. Since we already loaded Sentry (which is the main concern), we can allow it to continue being used for the remainder of the session! I think most of the simplification can come from this part. This would let you remove the changes from `blazepress-widget/index.tsx`, and you could greatly simplify `promote-post/index.ts`.

If we run into issues, we can adjust the sample rate rather than closing Sentry. I think it adds too much complexity, and I don't like how we could potentially be enabling or disabling Sentry multiple times per session.

My reasoning is that Sentry is designed to be globally enabled 100% of the time. Our async load stuff is already bespoke and adds a layer of complexity, but it's important for performance. But as much as possible, I want consumers of Sentry in Calypso to treat it how it was designed to be used, which is with the global configuration, and global methods like `captureException`. The cleanup stuff you needed to add for the widget is a bit too much extra complexity for consumers IMO.

sbarbosa committed:
Reduce complexity of the solution by avoiding closing the sentry instance on widget close

sbarbosa commented:
Thanks for the answer @noahtallen!
I applied the changes you proposed, the solution got simpler this way 👍

noahtallen commented on client/lib/promote-post/index.ts:
```diff
@@ -205,6 +213,12 @@ export async function showDSP(
 	} );
 }
 
+export async function cleanupDSP() {
```
Is this needed any more?

sbarbosa replied:
Yes, we still need to do some cleanups in the widget.

noahtallen reviewed (approved):
Thanks for the improvements! The code changes are looking good to me now.

timur987 reviewed (approved):
Thanks @sbarbosa!

The changes look good and work as advertised.

I wanted to highlight a couple of thoughts though.

1. Noah mentioned this https://github.com/Automattic/wp-calypso/pull/84564/files#r1425982797. I tried to find `cleanup` method in `window.BlazePress` but found only `render`. So it might not be needed.

2. I tried to follow scenario 1 you've mentioned in testing instructions but can't repeat the step `Verify that in the console, the reporting initialization text is shown. (We enable Sentry for all of the users that opens the widget.)`, I had nothing in my console after opening a widget (see video below for more details).

https://github.com/Automattic/wp-calypso/assets/115007291/95c3f673-252d-47fd-8e0e-bcdb08b8e829

Other than that everything works smoothly. Thanks!
Do not see any blockers for merging it.

sbarbosa commented:
Hi @timur987, thanks for the review!

>  I tried to find cleanup method in window.BlazePress but found only render

The widget you need to see is in the DSP PR 675. I made Calypso code compatible, so it won't break if the current widget doesn't have a cleanup method. But, when I merged this PR, and the widget counterpart, it will start to have.
We need that cleanup to restore the event processor we attached on the widget (it is the one in charge of tagging our errors).

> I tried to follow scenario 1 you've mentioned in testing instructions but can't repeat the step

Sorry, this is my fault. I recently undo the changes of the wp-calypso config file because they are not needed if you enable the flag from the URL. I updated the testing steps, but forgot to update the scenario 1.

What happened is that you need to have enabled the error reporting; without that flag, it won't report anything. What you need to do is to append  `?flags=catch-js-errors` to the URL. This will enable error reporting to 10% of the users; this way you will be able to test the PR correctly. I updated the testing steps to match this.

timur987 commented:
> I recently undo the changes of the wp-calypso config file because they are not needed if you enable the flag from the URL. I updated the testing steps, but forgot to update the scenario 1.

No worries, @sbarbosa, the only reason I mentioned that was to make sure we agree that the suggested change works properly. Since we did it, I don't have any questions/suggestions 🙂

sbarbosa committed:
Merge branch 'trunk' into update/promote-post-widget-cleanup-step

sbarbosa committed:
Restore changes done to the dashboard that generated a test error
